### DISCUSSION TITLE
Should we use `import.meta.url` to load `.data` packages when enabling ESM?

### DISCUSSION THREAD

**Discussion by guyutongxue** [emscripten-core/emscripten#21202] [Ideas]
When enabling `-sEXPORT_ES6`, we load `.wasm` relatively to `.js` automatically by using `import.meta.url` (if `Module.locateFile` not provided):

https://github.com/emscripten-core/emscripten/blob/ce5114bdd2175c7297583d3c25a53ca95d22f4ce/src/preamble.js#L626

But when loading `.data` files, the generated code just examines `Module.locateFile`, or else the raw file name is used:

https://github.com/emscripten-core/emscripten/blob/ce5114bdd2175c7297583d3c25a53ca95d22f4ce/tools/file_packager.py#L770

This is some kind of inconsistence... would it better to make `loadPackage` support `import.meta.url`?

**Comment by sbc100** [comment]
Sounds reasonable.  I guess that woulds mean some kind of new `--es6-module` setting for the file packager.

Can I ask, is there some reason why you don't use `--embed-file` though?

**Reply by guyutongxue** [reply]
> Can I ask, is there some reason why you don't use `--embed-file` though?   

Well I was building clang (clangd exactly) to wasm and it needs a complete sysroot (nearly 20MiB I guess) to run. And the wasm file itself is ~50MiB. So I think using `preload` maybe better than `embed`... just don't make a single file too big to fetch.

**Reply by sbc100** [reply]
I see. Perhaps that makes sense.  Remember though that embedding is more efficient as there is no copying needed.  With `--preload-file` you have to copy all of the data from the downloaded file into the linear memory.  So you might want to measure to see which is faster.
